ic-agent/src/agent/mod.rs

```diff
@@ -618,13 +618,29 @@ impl Agent {
         }
     }
 
-    fn check_delegation(&self, delegation: &Option<Delegation>) -> Result<Vec<u8>, AgentError> {
+    fn check_delegation(
+        &self,
+        delegation: &Option<Delegation>,
+        canister: &Principal,
+    ) -> Result<Vec<u8>, AgentError> {
         match delegation {
             None => self.read_root_key(),
             Some(delegation) => {
                 let cert: Certificate = serde_cbor::from_slice(&delegation.certificate)
                     .map_err(AgentError::InvalidCborData)?;
-                self.verify(&cert)?;
+                self.verify(&cert, canister)?;
+                let canister_range_lookup = [
+                    "subnet".into(),
+                    delegation.subnet_id.clone().into(),
+                    "canister_ranges".into(),
+                ];
+                let canister_range = lookup_value(&cert, canister_range_lookup)?;
+                let ranges: Vec<Vec<Principal>> =
```

Comment:
```suggestion
                let ranges: Vec<(Principal, Principal)> =
```
This way we won't crash the agent if the canister ranges are invalid (contain a single principal instead of a range).

Reply:
implemented